static/js/import_wizard/utils/detection.ts

```diff
@@ -0,0 +1,216 @@
+/**
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import _ from "lodash";
+
+import {
+  Column,
+  MappedThing,
+  Mapping,
+  MappingType,
+  MappingVal,
+  TypeProperty,
+} from "../types";
+
+const COLUMN_API_KEY = "column";
+const COLUMN_ID_API_KEY = "id";
+const COLUMN_HEADER_API_KEY = "header";
+const COLUMN_IDX_API_KEY = "column_idx";
+const DCID_API_KEY = "dcid";
+const DISP_NAME_API_KEY = "display_name";
+const PLACETYPE_API_KEY = "place_type";
+const PLACEPROP_API_KEY = "place_property";
+const DCTYPE_API_KEY = "dc_type";
+const DCPROP_API_KEY = "dc_property";
+const HEADER_API_KEY = "headers";
+
+function isDCTypeOrProperty(entity): boolean {
+  if (_.has(entity, DCID_API_KEY) && _.has(entity, DISP_NAME_API_KEY)) {
+    return true;
+  }
+  return false;
+}
+
+function isColumn(candidate): boolean {
+  if (
+    _.has(candidate, COLUMN_ID_API_KEY) &&
+    _.has(candidate, COLUMN_HEADER_API_KEY) &&
+    _.has(candidate, COLUMN_IDX_API_KEY)
+  ) {
+    return true;
+  }
+  return false;
+}
+
+/**
+ * Given the API response from the detection server API,
+ * parse it in to a Mapping structure.
+ *
+ * @param detectedResponse is the Object response from the server API.
+ *
+ * @returns a Mapping structure which has details of all the detected
+ *  columns. If there are any unexpected validation errors, that particular
+ *  MappingVal is skipped.
+ */
+export function parseDetectionAPIResponse(detectedResponse): Mapping {
+  const mToReturn: Mapping = new Map<MappedThing, MappingVal>();
+
+  // Some validations.
+  for (const [key, mVal] of Object.entries(detectedResponse)) {
+    // Keys must be among the MappedThing enum.
+    if (!_.includes(Object.values(MappedThing), key)) {
```

Comment:
also consider pulling out validation to a helper for maintainability. in general, shortening this function would be helpful, especially since this is such a long for loop with nested `continue` statements over 30 lines away from the `for` statement.

Reply:
Added a TODO for parsing helpers.